### PR TITLE
sssd functional test

### PR DESCRIPTION
**Pull request by tonyyuan1**
sssd functional test:
1. As openldap will be discarded from sles. sssd automation test will be split into two test modules like sssd_openldap and sssd_389ds.  When openldap being discarded we will just simply unschedule sssd_openldap and will not have to change code.
As there are already some sssd tests developed by some squads running in OSD, we could name the two modules sssd_openldap_functional, sssd_389ds_functional.
2. Setting up openldap/389-ds server in container and running the tests in single machine. It makes the automation development and maintenance much easier then multi-machines.
3. test cases and server configuration files are all placed in testopia
sssd openldap: https://bugzilla.suse.com/tr_show_case.cgi?case_id=1768711
sssd 389-ds: https://bugzilla.suse.com/tr_show_case.cgi?case_id=1768710

- Related ticket: https://progress.opensuse.org/issues/89479
- Needles: no
- Verification run: 
- sles15sp3: http://10.67.17.201/tests/1560
- sles15sp2: http://10.67.17.201/tests/1574
- sles15sp1: http://10.67.17.201/tests/1575
- sles12sp5: http://10.67.17.201/tests/1566
- opensuse: http://10.67.17.201/tests/1623
- opensuse: http://10.67.17.201/tests/1646
- ppc: https://openqa.suse.de/tests/6122315
- x390: https://openqa.suse.de/tests/6122308

### CODE REVIEW

**Comment by coolgw** [comment]
Talked with Tony , we need consider how this module integrate into migration group, so maybe extra lib/sssd.pm need created with function call, then migration process can easier call function check before / after migration.

**Comment by tonyyuan1** [comment]
> Talked with Tony , we need consider how this module integrate into migration group, so maybe extra lib/sssd.pm need created with function call, then migration process can easier call function check before / after migration.

migration test need to run sssd test twice in a single openqa job. Once before the migration, once after it. The second run uses the first sssd configuration files and active service so that it could check whether OS migration handles them correctly.
Solution: a global VAR,  the first run sets it to true so the second run skips configuration and jump to test case execution.

**Review by tjyrinki — APPROVED** [review]
Looking at the tests and verifications these look awesome from my side, but giving some time for others to possibly review it if there are still some things to consider before merging thist first version.

File: tests/console/sssd_openldap_functional.pm [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        # Configure sssd on the host
+        assert_script_run("curl -s " . data_url("sssd/openldap/sssd.conf") . " -o /etc/sssd/sssd.conf");
+        assert_script_run("curl -s " . data_url("sssd/openldap/nsswitch.conf") . " -o /etc/nsswitch.conf");
+        assert_script_run("curl -s " . data_url("sssd/openldap/ldapserver.crt") . " -o /etc/sssd/ldapserver.crt");
+        assert_script_run("curl -s " . data_url("sssd/openldap/config") . " --create-dirs -o ~/.ssh/config");
+        #    systemctl("stop nscd.service");
+        systemctl('disable --now nscd.service');
+
+        #    disable_and_stop_service("nscd");
```
Comment (dzedro) [inline tests/console/sssd_openldap_functional.pm:71]: disable_and_stop_service("nscd"); ?

File: tests/console/sssd_389ds_functional.pm [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    #offline identity lookup and authentification
+    assert_script_run("$docker stop ds389_container") if ($docker eq "docker");
+    #offline cached remote user indentity lookup
+    validate_script_output("id alice", sub { m/uid=9998\(alice\)/ });
+    #offline remote user authentification test
+    validate_script_output('sshpass -p open5use ssh mary@localhost whoami', sub { m/mary/ });
+    #offline sudo run a command as another user
+    validate_script_output('sshpass -p open5use ssh mary@localhost "echo open5use|sudo -S -u alice /usr/bin/cat /home/alice/hello"', sub { m/file read only by owner alice/ });
+}
+#rollback to clean up test enviroment
```
Comment (dzedro) [inline tests/console/sssd_389ds_functional.pm:109]: Roollback or not, if not then remove the commented lines

File: tests/console/sssd_openldap_functional.pm (hunk 1 above)
Comment (tonyyuan1) [reply tests/console/sssd_openldap_functional.pm]: yes, "enable --now" will do that, "stop and disable'.  It's one line code. disable_and_stop_service is multiple lines. So I am using "enable --now". I just undated code and removed the comments.

File: tests/console/sssd_389ds_functional.pm (hunk 2 above)
Comment (tonyyuan1) [reply tests/console/sssd_389ds_functional.pm]: No roollback, Just removed them.

File: tests/console/sssd_openldap_functional.pm (hunk 1 above)
Comment (dzedro) [reply tests/console/sssd_openldap_functional.pm]: I mean in comment you had stop and disable **nscd** but command was enabling **sssd**.

**Review by dzedro — APPROVED** [review]

File: tests/console/sssd_openldap_functional.pm (hunk 1 above)
Comment (tonyyuan1) [reply tests/console/sssd_openldap_functional.pm]: .  Line 68 and 71 were not comments.  They were source code which were commented out by me :). I forgot to deleted them.

**Comment by coolgw** [comment]
> > Talked with Tony , we need consider how this module integrate into migration group, so maybe extra lib/sssd.pm need created with function call, then migration process can easier call function check before / after migration.
> 
> migration test need to run sssd test twice in a single openqa job. Once before the migration, once after it. The second run uses the first sssd configuration files and active service so that it could check whether OS migration handles them correctly.
> Solution: a global VAR, the first run sets it to true so the second run skips configuration and jump to test case execution.

I proposal using implementation like:
https://confluence.suse.com/display/qasleapac1/Draft+service+check+design+and+implementation

File: tests/console/sssd_389ds_functional.pm [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+use base 'consoletest';
+use testapi;
+use strict;
+use warnings;
+use utils;
+use version_utils;
+use registration 'add_suseconnect_product';
+
+sub run {
+    #    my ($self) = @_;
```
Comment (coolgw) [inline tests/console/sssd_389ds_functional.pm:39]: need remove?

File: tests/console/sssd_openldap_functional.pm [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    #Change back password of remote user
+    assert_script_run('sshpass -p n0vell88 ssh bob@localhost \'echo -e "n0vell88\nopen5use\nopen5use" | passwd\'');
+    validate_script_output('sshpass -p open5use ssh bob@localhost echo "Password changed back!"', sub { m/Password changed back/ });
+    #offline identity lookup and authentification
+    assert_script_run('docker stop ldap_container');
+    #offline cached remote user indentity lookup
+    validate_script_output("id bob", sub { m/uid=5009\(bob\)/ });
+    #offline remote user authentification test
+    validate_script_output('sshpass -p open5use ssh adam@localhost whoami', sub { m/adam/ });
+    ##offline sudo run a command as another user
```
Comment (coolgw) [inline tests/console/sssd_openldap_functional.pm:98]: ## -> #

**Comment by coolgw** [comment]
For test scope, this module will whether support following base or not?
11sp4 for openldap
12sp3 for openldap
12sp4 for openldap

**Review by tjyrinki — CHANGES_REQUESTED** [review]
Could you please change according to coolgw's requests?

**Comment by coolgw** [comment]
Base current container design we have some limitation from migration view. Such as:
1) 12sp5 -> 15sp4, we have to using sssd_openldap_functional to check, but what happen if openldap package not exist on 15sp4 anymore?
2) if we running test on other platform such as 390, power, the docker still support this or not?
So best solution i think is using Mult-machine to check sssd, one single VM such as using windows server or SLE for server role.

**Comment by coolgw** [comment]
Any test result on other platform?  
Such as s390/aarch/power?  Since i have no idea whether docker can work or not.
This maybe block issue for sssd check. @tonyyuan1 @tjyrinki

**Comment by tonyyuan1** [comment]
> > > Talked with Tony , we need consider how this module integrate into migration group, so maybe extra lib/sssd.pm need created with function call, then migration process can easier call function check before / after migration.
> > 
> > 
> > migration test need to run sssd test twice in a single openqa job. Once before the migration, once after it. The second run uses the first sssd configuration files and active service so that it could check whether OS migration handles them correctly.
> > Solution: a global VAR, the first run sets it to true so the second run skips configuration and jump to test case execution.
> 
> I proposal using implementation like:
> https://confluence.suse.com/display/qasleapac1/Draft+service+check+design+and+implementation

This proposal actually cuts test into pieces and wrap them in functions. It will add a lot unnecessary code, complicates tests and populates library.  It increases the cost of maintenance and decrease flexibility and scalability.  

> > > Talked with Tony , we need consider how this module integrate into migration group, so maybe extra lib/sssd.pm need created with function call, then migration process can easier call function check before / after migration.
> > 
> > 
> > migration test need to run sssd test twice in a single openqa job. Once before the migration, once after it. The second run uses the first sssd configuration files and active service so that it could check whether OS migration handles them correctly.
> > Solution: a global VAR, the first run sets it to true so the second run skips configuration and jump to test case execution.
> 
> I proposal using implementation like:
> https://confluence.suse.com/display/qasleapac1/Draft+service+check+design+and+implementation

**Comment by tonyyuan1** [comment]
> > > Talked with Tony , we need consider how this module integrate into migration group, so maybe extra lib/sssd.pm need created with function call, then migration process can easier call function check before / after migration.
> > 
> > 
> > migration test need to run sssd test twice in a single openqa job. Once before the migration, once after it. The second run uses the first sssd configuration files and active service so that it could check whether OS migration handles them correctly.
> > Solution: a global VAR, the first run sets it to true so the second run skips configuration and jump to test case execution.
> 
> I proposal using implementation like:
> https://confluence.suse.com/display/qasleapac1/Draft+service+check+design+and+implementation

This proposal actually cuts test into pieces and wrap them in functions. It will add a lot unnecessary code, complicates tests and populates library.  It increases the cost of maintenance and decrease flexibility and scalability.

**Comment by tonyyuan1** [comment]
> For test scope, this module will whether support following base or not?
> 11sp4 for openldap
> 12sp3 for openldap
> 12sp4 for openldap 
> Also for 389ds, we only cover base >=15SP3?

Both openldap and 389ds should work from sle12GA onward. I did not test 11sp4 which is far out of maintenance.

**Comment by tonyyuan1** [comment]
> Any test result on other platform?
No, I did not have HW. 
> Such as s390/aarch/power? Since i have no idea whether docker can work or not.

It is supposed to work if they have docker and ldap. 
> This maybe block issue for sssd check. @tonyyuan1 @tjyrinki

**Comment by tonyyuan1** [comment]
> Base current container design we have some limitation from migration view. Such as:
> 
> 1. 12sp5 -> 15sp4, we have to using sssd_openldap_functional to check, but what happen if openldap package not exist on 15sp4 anymore?
> 2. if we running test on other platform such as 390, power, the docker still support this or not?
>    So best solution i think is using Mult-machine to check sssd, one single VM such as using windows server or SLE for server role.

This sssd test is designed for functional test and maintenance update test . It will not help migration test as I told you. I recommend you not to use it.

**Comment by coolgw** [comment]
> > For test scope, this module will whether support following base or not?
> > 11sp4 for openldap
> > 12sp3 for openldap
> > 12sp4 for openldap
> > Also for 389ds, we only cover base >=15SP3?
> 
> Both openldap and 389ds should work from sle12GA onward. I did not test 11sp4 which is far out of maintenance.
Could you help give some test result for 12sp3/4 ?

**Comment by coolgw** [comment]
> > For test scope, this module will whether support following base or not?
> > 11sp4 for openldap
> > 12sp3 for openldap
> > 12sp4 for openldap
> > Also for 389ds, we only cover base >=15SP3?
> 
> Both openldap and 389ds should work from sle12GA onward. I did not test 11sp4 which is far out of maintenance.

Could you help give some test result for 12sp3/4 ?

**Comment by coolgw** [comment]
> > Any test result on other platform?
> > No, I did not have HW.
> > Such as s390/aarch/power? Since i have no idea whether docker can work or not.
> 
> It is supposed to work if they have docker and ldap.
> 
> > This maybe block issue for sssd check. @tonyyuan1 @tjyrinki

could you help give some test result for other platform(s390/aarch/power)?

**Comment by tonyyuan1** [comment]
> > > For test scope, this module will whether support following base or not?
> > > 11sp4 for openldap
> > > 12sp3 for openldap
> > > 12sp4 for openldap
> > > Also for 389ds, we only cover base >=15SP3?
> > 
> > 
> > Both openldap and 389ds should work from sle12GA onward. I did not test 11sp4 which is far out of maintenance.
> 
> Could you help give some test result for 12sp3/4 ?

yes, sle12sp2: http://10.67.17.201/tests/1651

**Comment by tonyyuan1** [comment]
> > > Any test result on other platform?
> > > No, I did not have HW.
> > > Such as s390/aarch/power? Since i have no idea whether docker can work or not.
> > 
> > 
> > It is supposed to work if they have docker and ldap.
> > > This maybe block issue for sssd check. @tonyyuan1 @tjyrinki
> 
> could you help give some test result for other platform(s390/aarch/power)?

I can't because of no HW.

File: tests/console/sssd_389ds_functional.pm (hunk 3 above)
Comment (tonyyuan1) [reply tests/console/sssd_389ds_functional.pm]: Keep it as note that the test can run in serial console.

File: tests/console/sssd_openldap_functional.pm (hunk 4 above)
Comment (tonyyuan1) [reply tests/console/sssd_openldap_functional.pm]: fixed

**Comment by pdostal** [comment]
> > > > Any test result on other platform?
> > > > No, I did not have HW.
> > > > Such as s390/aarch/power? Since i have no idea whether docker can work or not.
> > > 
> > > 
> > > It is supposed to work if they have docker and ldap.
> > > > This maybe block issue for sssd check. @tonyyuan1 @tjyrinki
> > 
> > 
> > could you help give some test result for other platform(s390/aarch/power)?
> 
> I can't because of no HW.
You can use `openqa-clone-custom-git-refspec` against OSD and/or O3.

**Comment by tjyrinki** [comment]
OSD/O3 can be indeed used to test multi-arch, but usually PRs are fine with x86 because producing also those manually is time consuming work. If there's a problem with some other architectures, the test can be disabled there with a ticket filed.

SLE11 is not tested in openQA

There is a combination of factors like Tony switching to another squad, the ticket not mentioning migration (maybe it should have, but the new migration policy was a very new proposal at that point), and also that policy about the migration service check and implementation is still a draft despite accepted eg by me too, but many voices in QE LSG are unheard there currently.

The current PR solves a real problem described in the ticket, and for my part not due to any single reason but all of the above together I'd revert my earlier review and approve merging it, and filing a new ticket to modify it to migration needs, if the test is something migration would like to execute.

**Review by tjyrinki — APPROVED** [review]

**Comment by tonyyuan1** [comment]
@pdostal thanks!

> > > Any test result on other platform?
> > > No, I did not have HW.
> > > Such as s390/aarch/power? Since i have no idea whether docker can work or not.
> > 
> > 
> > It is supposed to work if they have docker and ldap.
> > > This maybe block issue for sssd check. @tonyyuan1 @tjyrinki
> 
> could you help give some test result for other platform(s390/aarch/power)?

Yes, see the VR part.

**Comment by coolgw** [comment]
> OSD/O3 can be indeed used to test multi-arch, but usually PRs are fine with x86 because producing also those manually is time consuming work. If there's a problem with some other architectures, the test can be disabled there with a ticket filed.
> 
> SLE11 is not tested in openQA
> 
> There is a combination of factors like Tony switching to another squad, the ticket not mentioning migration (maybe it should have, but the new migration policy was a very new proposal at that point), and also that policy about the migration service check and implementation is still a draft despite accepted eg by me too, but many voices in QE LSG are unheard there currently.
> 
> The current PR solves a real problem described in the ticket, and for my part not due to any single reason but all of the above together I'd revert my earlier review and approve merging it, and filing a new ticket to modify it to migration needs, if the test is something migration would like to execute.

Could you help notify me new ticket number once created ?  Thanks a lot.

**Pull request merged**